File: binary_search.cpp
// binary_search.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include <limits>
#include <sstream>
#include <iomanip>

struct DataRow {
    long key;
    std::string text;
};

bool readCSV(const std::string &fname, std::vector<DataRow> &out) {
    std::ifstream in(fname);
    if (!in) return false;
    std::string line;
    while (std::getline(in, line)) {
        auto sep = line.find(',');
        if (sep == std::string::npos) continue;
        try {
            long k = std::stol(line.substr(0, sep));
            out.push_back({k, line.substr(sep+1)});
        } catch (...) {}
    }
    return true;
}

int binarySearch(const std::vector<DataRow> &A, long target) {
    int lo = 0, hi = (int)A.size() - 1;
    while (lo <= hi) {
        int mid = lo + (hi - lo)/2;
        if (A[mid].key == target)      return mid;
        else if (A[mid].key < target)  lo = mid + 1;
        else                            hi = mid - 1;
    }
    return -1;
}

static std::string commafy(long long x) {
    std::ostringstream ss;
    ss.imbue(std::locale(""));
    ss << x;
    return ss.str();
}

int main(){
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::cout << "Enter path to sorted CSV file: ";
    std::string fname;
    if (!std::getline(std::cin, fname) || fname.empty()) {
        std::cerr << "No filename\n";
        return 1;
    }

    std::vector<DataRow> data;
    if (!readCSV(fname, data)) {
        std::cerr << "Error opening " << fname << "\n";
        return 1;
    }
    int N = (int)data.size();
    std::cout << "Loaded " << commafy(N) << " records from: " << fname << "\n\n";

    std::cout << "Running benchmark (" << commafy(N) << " searches)...\n";
    // Warm-up
    for (int i = 0; i < N; ++i)
        binarySearch(data, data[i].key);
    std::cout << "Warming up... done.\n\n";

    // Benchmark
    long long best_ns  = std::numeric_limits<long long>::max();
    long long worst_ns = 0;
    long long sum_ns   = 0;

    auto start_all = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < N; ++i) {
        auto t0 = std::chrono::high_resolution_clock::now();
        binarySearch(data, data[i].key);
        auto t1 = std::chrono::high_resolution_clock::now();
        long long ns = std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();
        sum_ns += ns;
        if (ns < best_ns)  best_ns  = ns;
        if (ns > worst_ns) worst_ns = ns;
    }
    auto end_all = std::chrono::high_resolution_clock::now();
    long long total_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(end_all - start_all).count();

    double best_ms    = best_ns  / 1e6;
    double worst_ms   = worst_ns / 1e6;
    double avg_ns     = double(sum_ns) / N;
    double avg_ms     = avg_ns   / 1e6;
    double total_ms   = total_ns / 1e6;
    double throughput = N / total_ms;  // searches per ms

    // Output
    std::cout << "\nPerformance Metrics:\n";
    std::cout << "--------------------------------\n";
    std::cout << "Dataset size:  " << commafy(N) << " records\n";
    std::cout << "Best case:     " << commafy(best_ns)  << " ns"
              << " (" << std::fixed << std::setprecision(6) << best_ms  << " ms)\n";
    std::cout << "Worst case:    " << commafy(worst_ns) << " ns"
              << " (" << std::fixed << std::setprecision(6) << worst_ms << " ms)\n";
    std::cout << "Average:       " << commafy((long long)avg_ns) << " ns"
              << " (" << std::fixed << std::setprecision(6) << avg_ms   << " ms)\n";
    std::cout << "--------------------------------\n";
    std::cout << "Total time:    " << commafy(total_ns) << " ns"
              << " (" << std::fixed << std::setprecision(3) << total_ms   << " ms)\n";
    std::cout << "Throughput:    " << std::fixed << std::setprecision(1)
              << throughput << " searches/ms\n";
    std::cout << "--------------------------------\n";

    return 0;
}

File: BinarySearch.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class BinarySearch {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter path to sorted CSV file: ");
        String filePath = scanner.nextLine().trim();

        try {
            // Read and validate the sorted CSV file
            List<Record> records = readCSV(filePath);
            System.out.println("\nLoaded " + records.size() + " records from: " + filePath);

            // Benchmark search performance
            benchmarkSearch(records);

        } catch (IOException e) {
            System.err.println("\nError: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }

    static class Record implements Comparable<Record> {
        final String id;
        final String value;

        Record(String id, String value) {
            this.id = id;
            this.value = value;
        }

        @Override
        public int compareTo(Record other) {
            return this.id.compareTo(other.id);
        }
    }

    private static List<Record> readCSV(String filePath) throws IOException {
        List<Record> records = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                if (parts.length == 2) {
                    records.add(new Record(parts[0].trim(), parts[1].trim()));
                }
            }
        }

        // Verify the records are sorted by ID
        for (int i = 1; i < records.size(); i++) {
            if (records.get(i-1).compareTo(records.get(i)) > 0) {
                throw new IOException("CSV is not sorted by ID. Problem at line " + (i+1) +
                        "\nFound: " + records.get(i-1).id +
                        " before " + records.get(i).id);
            }
        }

        return records;
    }

    private static Record binarySearch(List<Record> records, String targetId) {
        int left = 0;
        int right = records.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            Record current = records.get(mid);
            int comparison = current.id.compareTo(targetId);

            if (comparison == 0) {
                return current;
            } else if (comparison < 0) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return null;
    }

    private static void benchmarkSearch(List<Record> records) {
        long minTime = Long.MAX_VALUE;
        long maxTime = Long.MIN_VALUE;
        long totalTime = 0;
        int totalSearches = records.size();

        System.out.println("\nRunning benchmark (" + totalSearches + " searches)...");

        // Warm up the JVM
        System.out.print("Warming up...");
        for (int i = 0; i < Math.min(1000, records.size()); i++) {
            binarySearch(records, records.get(i).id);
        }
        System.out.println(" done.");

        // Main benchmarking
        for (Record record : records) {
            long startTime = System.nanoTime();
            Record found = binarySearch(records, record.id);
            long endTime = System.nanoTime();
            long duration = endTime - startTime;

            if (found == null || !found.id.equals(record.id)) {
                System.err.println("Search failed for ID: " + record.id);
                continue;
            }

            minTime = Math.min(minTime, duration);
            maxTime = Math.max(maxTime, duration);
            totalTime += duration;
        }

        // Calculate statistics
        double avgTime = (double) totalTime / totalSearches;
        double minMs = minTime / 1_000_000.0;
        double maxMs = maxTime / 1_000_000.0;
        double avgMs = avgTime / 1_000_000.0;

        // Display performance metrics
        System.out.println("\nPerformance Metrics:");
        System.out.println("--------------------------------");
        System.out.printf("Dataset size:    %,d records\n", totalSearches);
        System.out.printf("Best case:       %,d ns (%.6f ms)\n", minTime, minMs);
        System.out.printf("Worst case:      %,d ns (%.6f ms)\n", maxTime, maxMs);
        System.out.printf("Average:         %,.0f ns (%.6f ms)\n", avgTime, avgMs);
        System.out.println("--------------------------------");
        System.out.printf("Total time:      %,d ns (%.3f ms)\n", totalTime, totalTime / 1_000_000.0);
        System.out.printf("Throughput:      %,.1f searches/ms\n", totalSearches / (totalTime / 1_000_000.0));
        System.out.println("--------------------------------");
    }
}